### PR TITLE
ads: change xDS path responses to generic interface

### PR DESCRIPTION
**Pull request by eduser25**
Currently, each xDS path will marshal and create each DiscoveryResponses
on the same vertical. This limits the vision that ads/server has over responses,
and and in addition causes the marshalling logic to exist in all paths.

This commit decouples getting the xDS resources from the Marshal/SendToServer
logic.

This removes some common code for all xDS paths, but more
importantly enable us to unlock additional features such as global consistency
checks between Requests/Responses, use of xDS server cache implementations,
etc.

Signed-off-by: Eduard Serra <eduser25@gmail.com>

- Control Plane          [x]

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

File: pkg/envoy/ads/response.go [hunk 1]
```diff
@@ -107,14 +109,27 @@ func (s *Server) newAggregatedDiscoveryResponse(proxy *envoy.Proxy, request *xds
...
 	if err != nil {
 		log.Error().Msgf("Responder for TypeUrl %s is not implemented", request.TypeUrl)
 		return nil, errCreatingResponse
 	}
 
-	response.Nonce = proxy.SetNewNonce(typeURL)
-	response.VersionInfo = strconv.FormatUint(proxy.IncrementLastSentVersion(typeURL), 10)
+	response := &xds_discovery.DiscoveryResponse{
+		TypeUrl:     request.TypeUrl, // Request TypeURL
+		Resources:   []*anypb.Any{},
```
Comment (shashankram) [inline pkg/envoy/ads/response.go:120]: Not needed.

**Review by shashankram — APPROVED** [review]

File: pkg/envoy/ads/server.go [hunk 2]
```diff
@@ -26,7 +27,7 @@ const ServerType = "ADS"
 func NewADSServer(meshCatalog catalog.MeshCataloger, enableDebug bool, osmNamespace string, cfg configurator.Configurator, certManager certificate.Manager) *Server {
 	server := Server{
 		catalog: meshCatalog,
-		xdsHandlers: map[envoy.TypeURI]func(catalog.MeshCataloger, *envoy.Proxy, *xds_discovery.DiscoveryRequest, configurator.Configurator, certificate.Manager) (*xds_discovery.DiscoveryResponse, error){
+		xdsHandlers: map[envoy.TypeURI]func(catalog.MeshCataloger, *envoy.Proxy, *xds_discovery.DiscoveryRequest, configurator.Configurator, certificate.Manager) ([]types.Resource, error){
```
Comment (draychev) [inline pkg/envoy/ads/server.go:30]: I worry about the ambiguity of `[]types.Resource`. At first glance I learned that:
 - we have a list of things
 - they are in a package called types
 - they are a resource

Should this be something like `[]xds_cache.DiscoveryResponse`?  Is there a better name to increase the readability here?

File: pkg/envoy/cds/response.go [hunk 3]
```diff
@@ -68,26 +68,17 @@ func NewResponse(meshCatalog catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_d
...
-			log.Error().Err(err).Msgf("Failed to marshal cluster %s for Envoy with XDS Certificate SerialNumber=%s on Pod with UID=%s",
-				cluster.Name, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-			return nil, err
-		}
-		resp.Resources = append(resp.Resources, marshalledClusters)
+		resources = append(resources, cluster)
 	}
 
-	return resp, nil
+	return resources, nil
```
Comment (draychev) [inline pkg/envoy/cds/response.go:83]: This is also really tough to understand for a newcomer.
I see that the return type of the function is `[]type.Resource` and the function returns `resources`.
So this is great for the compiler, but I still don't know what this is.
I see that we are adding `cluster` to this - and from that I can start to understand that this is a collection of most likely CDS responses.
Is there a way to give more semantics to the slice name?
Because the function here is `cds.NewResponse()`, could this be `clusters` or `cdsResponses` or `xdsClusters`?
My intent is to make this easier to read - how do we stuff more context in this one line so one does not need to read the rest of the function to understand what is being returned.

File: pkg/envoy/ads/server.go (hunk 2 above)
Comment (eduser25) [reply pkg/envoy/ads/server.go:30]: Resources is, even by checking envoy xDS proto docs and argot, a common name to refer to the object contents of the response for a given `TypeURI`. Each Response contains a number of Resources. 

I think the misunderstanding is that the handler no longer returns the marshalled objects in a prefilled DiscoveryResponse, but instead the list of Resources related to the TypeURI the handler will handle.

If this is a clarification problem, I can further comment the code or even typify the map and function signature to express additional information.
If you think this is otherwise a design problem, let's talk offline as we could have different views of what this code should/could do.

File: pkg/envoy/cds/response.go (hunk 3 above)
Comment (eduser25) [reply pkg/envoy/cds/response.go:83]: I can definitely try to add more context on the response variables if you think current code presents readability problems.

File: pkg/envoy/ads/types.go [hunk 4]
```diff
@@ -18,10 +19,18 @@ var (
 	log = logger.New("envoy/ads")
 )
 
+// XdsResponseFunction is a function type entry point for xDS verticals.
+// Params:
+// - It gets OSM configuration state provided by catalog, configurator and certificate manager interfaces
+// - It gets references to the Proxy and the associated discovery request that triggered the execution of the given function (segregated by TypeURI)
+// Returns:
+// - Returns a list of resources composed by the related request and TypeURL call.
+type XdsResponseFunction func(catalog.MeshCataloger, *envoy.Proxy, *xds_discovery.DiscoveryRequest, configurator.Configurator, certificate.Manager) ([]types.Resource, error)
```
Comment (draychev) [inline pkg/envoy/ads/types.go:28]: It seems to me that this is only used within `pkg/envoy/ads` so we don't need to export it (so lower case xds could be better).

Does it make sense to also pull this type in a separate PR since it seems orthogonal to some of the other changes in this PR?

Comment (eduser25) [reply pkg/envoy/ads/types.go:28]: yes, absolutely. I'll push it separately.

**Review by michelleN — APPROVED** [review]
Thanks for the cleanup and refactor @eduser25. I appreciate that we're making the code more generic here. It is not easy understanding what `xds_discovery.DiscoveryResponse` is as a newcomer so this is a step in the right direction. I also feel what @draychev described about types.Resources being slightly confusing too. I agree we need to work to better document in the code either at the entry point and/or at each signature what resources is supposed to be.
